src/jit.c

```diff
@@ -0,0 +1,1550 @@
+ * despite being non-volatile.
+ */
+#if defined(_WIN32)
+static const int nonvolatile_reg[] = {RBP, RBX, RDI, RSI, R13, R14, R15};
+static const int parameter_reg[] = {RCX, RDX, R8, R9};
+#define RCX_ALT R10
+static const int register_map[REGISTER_MAP_SIZE] = {
+    RAX, R10, RDX, R8, R9, R14, R15, RDI, RSI, RBX, RBP,
+};
+#else
+#define RCX_ALT R9
+static const int nonvolatile_reg[] = {RBP, RBX, R13, R14, R15};
+static const int parameter_reg[] = {RDI, RSI, RDX, RCX, R8, R9};
+static const int temp_reg[] = {RAX, RBX, RCX};
+static const int register_map[REGISTER_MAP_SIZE] = {
+    RAX, RDI, RSI, RDX, R9, R8, RBX, R13, R14, R15, RBP,
+};
+#endif
+#elif defined(__aarch64__)
+/*
+ * There are two common x86-64 calling conventions, as discussed at
+ * https://en.wikipedia.org/wiki/X86_calling_conventions#x86-64_calling_conventions
+ *
+ * Please Note: R12 is special and we are *not* using it. As a result, it is
+ * omitted from the list of non-volatile registers for both platforms (even
+ * though it is, in fact, non-volatile).
+ *
+ * BPF R0-R4 are "volatile"
+ * BPF R5-R10 are "non-volatile"
+ * In general, we attempt to map BPF volatile registers to x64 volatile and BPF
+ * non- volatile to x64 non-volatile.
+ */
+
+/* callee_reg - this must be a multiple of two because of how we save the stack
+ * later on. */
+static const int callee_reg[] = {R19, R20, R21, R22, R23, R24, R25, R26};
+/* parameter_reg (Caller saved registers) */
+static const int parameter_reg[] = {R0, R1, R2, R3, R4};
+static const int temp_reg[] = {R6, R7, R8};
+/* Temp register for immediate generation */
+static const int temp_imm_reg = R24;
+/* Temp register for division results */
+static const int temp_div_reg = R25;
+
+/*  Register assignments:
+ *  Arm64       Usage
+ *   r0 - r4     Function parameters, caller-saved
+ *   r6 - r8     Temp - used for storing calculated value during execution
+ *   r19 - r23   Callee-saved registers
+ *   r24         Temp - used for generating 32-bit immediates
+ *   r25         Temp - used for modulous calculations
+ */
+
+static const int register_map[REGISTER_MAP_SIZE] = {
+    R5,                      /* result */
+    R0,  R1,  R2,  R3,  R4,  /* parameters */
+    R19, R20, R21, R22, R23, /* callee-saved */
+};
+static inline void emit_load_imm(struct jit_state *state, int dst, int64_t imm);
+#endif
+
+/* Return the x86 register for the given JIT register */
+static int map_register(int r)
+{
+    assert(r < N_VM_REGS);
+    return register_map[r % N_VM_REGS];
+}
+
+static inline void offset_map_insert(struct jit_state *state, int32_t target_pc)
+{
+    struct offset_map *map_entry = &state->offset_map[state->num_insn++];
+    map_entry->PC = target_pc;
+    map_entry->offset = state->offset;
+}
+
+#if !defined(__APPLE__)
+#define sys_icache_invalidate(addr, size) \
+    __builtin___clear_cache((char *) (addr), (char *) (addr) + (size));
+#endif
+
+static inline void emit_bytes(struct jit_state *state, void *data, uint32_t len)
```

Comment:
Don't inline this function. Let compiler decide the optimizations when LTO is enabled.